### ISSUE TITLE
Regularisation losses with nested layers and BasicDecoder

### ISSUE THREAD

**Issue by georgesterpu** [tensorflow/addons#577]
**System information**
- OS Platform and Distribution : Manjaro Linux testing
- TensorFlow version: tensorflow-gpu 2.0 from pypi
- TensorFlow-Addons version: 0.6.0 from pypi
- Python version: 3.7.4
- Is GPU used? (yes/no): yes

**Describe the bug**

Building a seq2seq model based on tfa.seq2seq seems to return a wrong number of regularisation losses for the decoder. Assume that both the encoder and the decoder use LSTM cells with the `kernel_regularizer` argument set. No other variables were regularised in the making of this model.
Calling `model.losses` (where model is an instance of `tf.keras.Model`) appears to return the correct number of losses for the Encoder module (one regularisation loss per LSTM layer), but the incorrect one for the Decoder. In particular, it is the same loss value replicated three times, instead of appearing only once. The effect is an overestimation of the total batch loss, possibly causing divergence issues.

**Code to reproduce the issue**

I have implemented the training loop in the most basic way, like this:
```
    def _train_step(self, data):
        with tf.GradientTape() as tape:
            prediction = self.model(data, training=True)
            main_loss = loss_fn(logits=prediction, label_ids=data.labels, sequence_lengths=data.labels_length)
            loss = tf.add_n([main_loss] + self.model.losses)
```
and the model is implemented as in the readme file.

This is how `nested_layers` list within `self._gather_children_attribute('losses')` in `keras.engine.base_layer.py` looks like:

```
result = {list} <class 'list'>: [<tensorflow.python.keras.layers.recurrent_v2.LSTMCell object at 0x7f80502bd9d0>, <tensorflow.python.keras.layers.core.Dense object at 0x7f80502bd2d0>, <tensorflow.python.keras.layers.embeddings.Embedding object at 0x7f80502a4c10>, <tensorf
 0 = {LSTMCell} <tensorflow.python.keras.layers.recurrent_v2.LSTMCell object at 0x7f80502bd9d0>
 1 = {Dense} <tensorflow.python.keras.layers.core.Dense object at 0x7f80502bd2d0>
 2 = {Embedding} <tensorflow.python.keras.layers.embeddings.Embedding object at 0x7f80502a4c10>
 3 = {BasicDecoder} <tensorflow_addons.seq2seq.basic_decoder.BasicDecoder object at 0x7f80503946d0>
 4 = {BeamSearchDecoder} <tensorflow_addons.seq2seq.beam_search_decoder.BeamSearchDecoder object at 0x7f80502a4410>
 __len__ = {int} 5
```

The issue here is that `LSTMCell`, `BasicDecoder`, and `BeamSearchDecoder` all return the same `.losses` value. All three classes are instantiated in my Decoder's constructor. It looks like the two decoders need to be removed from the list, since they only hold a reference to `LSTMCell`, while the variables were not created there. I suspect this behaviour may also appear in other similar cases where there is a "has" relationship between classes.

**Other info / logs**

pinging @fchollet as well since the git log indicates familiarity with the code in `base_layer.py`.

**Comment by georgesterpu** [comment]
Here is a minimal example to reproduce the issue outside the scope of tfa.seq2seq
```
import tensorflow as tf


class A(tf.keras.layers.Layer):
    def __init__(self, layer):
        super(A, self).__init__()
        self.layer = layer

    def call(self, inputs):
        return self.layer(inputs)


class B(tf.keras.layers.Layer):
    def __init__(self):
        super(B, self).__init__()
        self.obj = tf.keras.layers.Dense(13, kernel_regularizer=tf.keras.regularizers.l1(5))
        self.layerB = A(self.obj)

    def call(self, inputs):
        return self.layerB(inputs)

model = B()

output = model(tf.ones([5, 10]))
print(len(model.losses))
```

Once we rename `self.obj` to `obj`, i.e. not saving it as a class member, we end up with a single regularisation loss.

**Comment by guillaumekln** [comment]
I believe this should be fixed in TensorFlow. For attributes such as `weights` or `trainable_weights`, there is already an explicit deduplication of the return variables (see the related commit https://github.com/tensorflow/tensorflow/commit/596d3ccb0303f54f6184f027840c69e7a6ca2b68). A similar logic should probably be applied for `losses`.

I would suggest opening an issue on the TensorFlow main repository with the code snippet above.

**Comment by seanpmorgan** [comment]
Closing now that an issue is in TF-Core. Thanks for reporting!
